### ISSUE TITLE
k hot encoding for tf.keras.preprocessing.image.DirectoryIterator

### ISSUE THREAD

**Issue by Teut2711** [tensorflow/tensorflow#43634]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

tf.keras.preprocessing.image.DirectoryIterator

**System information**
- TensorFlow version (you are using):latest
- Are you willing to contribute it (Yes/No):Yes

**Describe the feature and the current behavior/state.**
Currently the function `tf.keras.preprocessing.image.DirectoryIterator` doesn't have any option for multiple classes for a single image. Currently only a single class can be assigned to an image.

**Will this change the current api? How?**

**Who will benefit with this feature?**
Anyone

**Any Other info.**

**Comment by tanzhenyu** [comment]
This class is not meant to support k hot encoding, given it requires the directory to be split by class

**Comment by XtremeGood** [comment]
I 'm not able to understand why it can't be done. The subdirs can have same images and they will have same name. In case of k-hot encoding, all that's to be done to create a n-dim vector for each entry.
Example if,
A, B, C and D are the subdirs and A and B have a common file abc.png then corresponding to "abc.png" a label of array([1, 1, 0, 0]) can be made. I 'm unable to see what problems can be faced.

**Comment by XtremeGood** [comment]
Also, I w'd like to contribute if I can get a little bit of support because I haven't contributed so far yet.

**Comment by tanzhenyu** [comment]
> Also, I w'd like to contribute if I can get a little bit of support because I haven't contributed so far yet.

How would you like to propose it be done?

**Comment by XtremeGood** [comment]
> 
> 
> I 'm not able to understand why it can't be done. The subdirs can have same images and they will have same name. In case of k-hot encoding, all that's to be done is to create is a n-dim vector for each entry.
> Example if,
> A, B, C and D are the subdirs and A and B have a common file abc.png then corresponding to "abc.png" a label of array([1, 1, 0, 0]) can be made. I 'm unable to see what problems can be faced.

@tanzhenyu  As said here.

**Comment by tanzhenyu** [comment]
> > I 'm not able to understand why it can't be done. The subdirs can have same images and they will have same name. In case of k-hot encoding, all that's to be done is to create is a n-dim vector for each entry.
> > Example if,
> > A, B, C and D are the subdirs and A and B have a common file abc.png then corresponding to "abc.png" a label of array([1, 1, 0, 0]) can be made. I 'm unable to see what problems can be faced.
> 
> @tanzhenyu As said here.

That seems do-able, I was thinking something else. 
I will mark this as contribution welcome

**Comment by XtremeGood** [comment]
Thank you.

**Comment by imrahul361** [comment]
I would like to know how to approach this problem? and contribute to it. Please guide me

**Comment by XtremeGood** [comment]
I just started with the dev but I m unable to perform `pip install -e ./tensorflow'. 
```

C:\Users\Dell\Desktop\tf dev>"c:/Users/Dell/Desktop/tf dev/venv/Scripts/activate.bat"

(venv) C:\Users\Dell\Desktop\tf dev>pip install tensorflow
Collecting tensorflow
  Downloading tensorflow-2.3.1-cp38-cp38-win_amd64.whl (342.5 MB)
     |████████████████████████████████| 342.5 MB 1.3 kB/s
Collecting tensorboard<3,>=2.3.0
  Downloading tensorboard-2.3.0-py3-none-any.whl (6.8 MB)
     |████████████████████████████████| 6.8 MB 2.6 kB/s
Collecting absl-py>=0.7.0
  Downloading absl_py-0.10.0-py3-none-any.whl (127 kB)
     |████████████████████████████████| 127 kB 2.2 MB/s
Collecting numpy<1.19.0,>=1.16.0
  Downloading numpy-1.18.5-cp38-cp38-win_amd64.whl (12.8 MB)
     |████████████████████████████████| 12.8 MB 17 kB/s
Processing c:\users\dell\appdata\local\pip\cache\wheels\5f\fd\9e\b6cf5890494cb8ef0b5eaff72e5d55a70fb56316007d6dfe73\wrapt-1.12.1-cp38-cp38-win_amd64.whl
Requirement already satisfied: wheel>=0.26 in c:\users\dell\desktop\tf dev\venv\lib\site-packages (from tensorflow) (0.35.1)
Collecting astunparse==1.6.3
  Downloading astunparse-1.6.3-py2.py3-none-any.whl (12 kB)
Collecting keras-preprocessing<1.2,>=1.1.1
  Downloading Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)
     |████████████████████████████████| 42 kB 57 kB/s
Collecting gast==0.3.3
  Downloading gast-0.3.3-py2.py3-none-any.whl (9.7 kB)
Collecting tensorflow-estimator<2.4.0,>=2.3.0
  Downloading tensorflow_estimator-2.3.0-py2.py3-none-any.whl (459 kB)
     |████████████████████████████████| 459 kB 2.2 MB/s
Collecting termcolor>=1.1.0
  Downloading termcolor-1.1.0.tar.gz (3.9 kB)
Collecting google-pasta>=0.1.8
  Downloading google_pasta-0.2.0-py3-none-any.whl (57 kB)
     |████████████████████████████████| 57 kB 404 kB/s
Collecting grpcio>=1.8.6
  Downloading grpcio-1.33.1-cp38-cp38-win_amd64.whl (2.7 MB)
     |████████████████████████████████| 2.7 MB 96 kB/s
Collecting h5py<2.11.0,>=2.10.0
  Downloading h5py-2.10.0-cp38-cp38-win_amd64.whl (2.5 MB)
     |████████████████████████████████| 2.5 MB 1.3 MB/s
Collecting opt-einsum>=2.3.2
  Downloading opt_einsum-3.3.0-py3-none-any.whl (65 kB)
     |████████████████████████████████| 65 kB 231 kB/s
Collecting six>=1.12.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting protobuf>=3.9.2
  Using cached protobuf-3.13.0-py2.py3-none-any.whl (438 kB)
Collecting werkzeug>=0.11.15
  Using cached Werkzeug-1.0.1-py2.py3-none-any.whl (298 kB)
Collecting google-auth<2,>=1.6.3
  Downloading google_auth-1.22.1-py2.py3-none-any.whl (114 kB)
     |████████████████████████████████| 114 kB 6.4 MB/s
Collecting requests<3,>=2.21.0
  Using cached requests-2.24.0-py2.py3-none-any.whl (61 kB)
Collecting tensorboard-plugin-wit>=1.6.0
  Downloading tensorboard_plugin_wit-1.7.0-py3-none-any.whl (779 kB)
     |████████████████████████████████| 779 kB 1.6 MB/s
Collecting google-auth-oauthlib<0.5,>=0.4.1
  Using cached google_auth_oauthlib-0.4.1-py2.py3-none-any.whl (18 kB)
Collecting markdown>=2.6.8
  Downloading Markdown-3.3.2-py3-none-any.whl (95 kB)
     |████████████████████████████████| 95 kB 312 kB/s
Requirement already satisfied: setuptools>=41.0.0 in c:\users\dell\desktop\tf dev\venv\lib\site-packages (from tensorboard<3,>=2.3.0->tensorflow) (50.3.2)
Collecting pyasn1-modules>=0.2.1
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting rsa<5,>=3.1.4; python_version >= "3.5"
  Using cached rsa-4.6-py3-none-any.whl (47 kB)
Collecting cachetools<5.0,>=2.0.0
  Using cached cachetools-4.1.1-py3-none-any.whl (10 kB)
Collecting idna<3,>=2.5
  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2020.6.20-py2.py3-none-any.whl (156 kB)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1
  Downloading urllib3-1.25.11-py2.py3-none-any.whl (127 kB)
     |████████████████████████████████| 127 kB 3.3 MB/s
Collecting chardet<4,>=3.0.2
  Using cached chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Collecting requests-oauthlib>=0.7.0
  Using cached requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Collecting oauthlib>=3.0.0
  Using cached oauthlib-3.1.0-py2.py3-none-any.whl (147 kB)
Building wheels for collected packages: termcolor
  Building wheel for termcolor (setup.py) ... done
  Created wheel for termcolor: filename=termcolor-1.1.0-py3-none-any.whl size=4835 sha256=db7b0ef9c0b0fd1eb0815ef56c712696cbfdac549f7da0d48335dc8efc4f5dd6    
  Stored in directory: c:\users\dell\appdata\local\pip\cache\wheels\a0\16\9c\5473df82468f958445479c59e784896fa24f4a5fc024b0f501
Successfully built termcolor
Installing collected packages: werkzeug, numpy, six, pyasn1, pyasn1-modules, rsa, cachetools, google-auth, idna, certifi, urllib3, chardet, requests, tensorboard-plugin-wit, grpcio, oauthlib, requests-oauthlib, google-auth-oauthlib, protobuf, markdown, absl-py, tensorboard, wrapt, astunparse, keras-preprocessing, gast, tensorflow-estimator, termcolor, google-pasta, h5py, opt-einsum, tensorflow
Successfully installed absl-py-0.10.0 astunparse-1.6.3 cachetools-4.1.1 certifi-2020.6.20 chardet-3.0.4 gast-0.3.3 google-auth-1.22.1 google-auth-oauthlib-0.4.1 google-pasta-0.2.0 grpcio-1.33.1 h5py-2.10.0 idna-2.10 keras-preprocessing-1.1.2 markdown-3.3.2 numpy-1.18.5 oauthlib-3.1.0 opt-einsum-3.3.0 protobuf-3.13.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.24.0 requests-oauthlib-1.3.0 rsa-4.6 six-1.15.0 tensorboard-2.3.0 tensorboard-plugin-wit-1.7.0 tensorflow-2.3.1 tensorflow-estimator-2.3.0 termcolor-1.1.0 urllib3-1.25.11 werkzeug-1.0.1 wrapt-1.12.1

(venv) C:\Users\Dell\Desktop\tf dev>pip install pytest
Collecting pytest
  Using cached pytest-6.1.1-py3-none-any.whl (272 kB)
Collecting colorama; sys_platform == "win32"
  Downloading colorama-0.4.4-py2.py3-none-any.whl (16 kB)
Collecting py>=1.8.2
  Using cached py-1.9.0-py2.py3-none-any.whl (99 kB)
Collecting packaging
  Using cached packaging-20.4-py2.py3-none-any.whl (37 kB)
Collecting pluggy<1.0,>=0.12
  Using cached pluggy-0.13.1-py2.py3-none-any.whl (18 kB)
Collecting attrs>=17.4.0
  Using cached attrs-20.2.0-py2.py3-none-any.whl (48 kB)
Collecting toml
  Using cached toml-0.10.1-py2.py3-none-any.whl (19 kB)
Collecting iniconfig
  Downloading iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Collecting atomicwrites>=1.0; sys_platform == "win32"
  Using cached atomicwrites-1.4.0-py2.py3-none-any.whl (6.8 kB)
Requirement already satisfied: six in c:\users\dell\desktop\tf dev\venv\lib\site-packages (from packaging->pytest) (1.15.0)
Collecting pyparsing>=2.0.2
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Installing collected packages: colorama, py, pyparsing, packaging, pluggy, attrs, toml, iniconfig, atomicwrites, pytest
Successfully installed atomicwrites-1.4.0 attrs-20.2.0 colorama-0.4.4 iniconfig-1.1.1 packaging-20.4 pluggy-0.13.1 py-1.9.0 pyparsing-2.4.7 pytest-6.1.1 toml-0.10.1

(venv) C:\Users\Dell\Desktop\tf dev>git clone https://github.com/tensorflow/tensorflow.git
Cloning into 'tensorflow'...
remote: Enumerating objects: 111, done.
remote: Counting objects: 100% (111/111), done.
remote: Compressing objects: 100% (86/86), done.
remote: Total 1023115 (delta 51), reused 46 (delta 25), pack-reused 1023004R
Receiving objects: 100% (1023115/1023115), 611.65 MiB | 3.42 MiB/s, done.
Resolving deltas: 100% (834072/834072), done.
Updating files: 100% (23593/23593), done.

(venv) C:\Users\Dell\Desktop\tf dev>pip install -e tensorflow
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: C:\Users\Dell\Desktop\tf dev\tensorflow

(venv) C:\Users\Dell\Desktop\tf dev>pip install -e ./tensorflow
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: C:\Users\Dell\Desktop\tf dev\tensorflow

(venv) C:\Users\Dell\Desktop\tf dev>pip install -e ./tensorflow/tensorflow
ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: C:\Users\Dell\Desktop\tf dev\tensorflow\tensorflow

Can I get some help here?

```

**Comment by XtremeGood** [comment]
@tanzhenyu I was starting with making it but it felt like `https://keras.io/api/preprocessing/image/#flowfromdataframe-method` might be the most suitable based on data that one gets on Kaggle. Please share your opinions.

**Comment by AdityaKane2001** [comment]
@tanzhenyu 
I want to propose a possible solution. Suppose following is the directory structure
```
images
-feature1
---img1.jpg
---img2.jpg
---img3.jpg
-feature2
---img1.jpg
---img3.jpg
---img4.jpg
---img5.jpg
```
We could calculate the hash of all images in all directories, as the names can be the same, and store them in a dict (say).
Structure of this dict:
```python
{
  'feature1':{
                    'img1.jpg':1234,
                    'img2.jpg':1235
                     ....  },
  'feature2':{
                   'img1.jpg':1234,
                    'img3.jpg':1234
                     ....  },
}

```
 Then make a `tf.dataDataset` of all unique images (no labels in that dataset). Then  iterate through all images in all feature dicts to find if it is present under any other feature, and create a many-hot vector in this fashion. Of course, this is a brute-force method. Please suggest improvements on the same.

**Comment by XtremeGood** [comment]
Actually keras provides already all required methods so it  would be just better to wrap them.

**Comment by AdityaKane2001** [comment]
@XtremeGood 
Even for this?

**Comment by XtremeGood** [comment]
keras provide total support so I don't think reinventing the wheel would do it any good.

**Comment by XtremeGood** [comment]
Actually no, keras does not provide enough good api like tensorflow, it rather requires you to load all data in memory together. So an efficient approach like storing the paths in a dataframe and making the iterator again and again would be required.

**Comment by AdityaKane2001** [comment]
I was thinking on similar lines, using `tf.data.Dataset` more efficiently

**Comment by AdityaKane2001** [comment]
sklearn has MultiLabelBinarizer. Intending to do something similar
